// OsirisCare Compliance Agent gRPC Service
//
// This defines the communication protocol between Go Agents on Windows
// workstations and the Python Compliance Agent on the NixOS appliance.

syntax = "proto3";

package compliance;

option go_package = "github.com/osiriscare/agent/proto";

// ComplianceAgent service handles communication from Go agents
service ComplianceAgent {
  // Register a new agent with the appliance
  rpc Register(RegisterRequest) returns (RegisterResponse);

  // Report drift events (streaming from agent)
  rpc ReportDrift(stream DriftEvent) returns (stream DriftAck);

  // Report healing results
  rpc ReportHealing(HealingResult) returns (HealingAck);

  // Send heartbeat
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);

  // Report RMM detection status
  rpc ReportRMMStatus(RMMStatusReport) returns (RMMAck);
}

// RegisterRequest is sent when an agent first connects
message RegisterRequest {
  string hostname = 1;
  string os_version = 2;
  string agent_version = 3;
  repeated string installed_software = 4;
  string mac_address = 5;
}

// RegisterResponse provides agent configuration
message RegisterResponse {
  string agent_id = 1;
  int32 check_interval_seconds = 2;
  repeated string enabled_checks = 3;
  CapabilityTier capability_tier = 4;
  map<string, string> check_config = 5;
}

// CapabilityTier defines what actions the agent can take
enum CapabilityTier {
  MONITOR_ONLY = 0;     // Can only report drift
  SELF_HEAL = 1;        // Can attempt automated remediation
  FULL_REMEDIATION = 2; // Can perform all remediations
}

// DriftEvent represents a compliance check result
message DriftEvent {
  string agent_id = 1;
  string hostname = 2;
  string check_type = 3;      // bitlocker, defender, firewall, patches, screenlock, rmm
  bool passed = 4;
  string expected = 5;
  string actual = 6;
  string hipaa_control = 7;   // e.g., "164.312(a)(2)(iv)"
  int64 timestamp = 8;        // Unix timestamp
  map<string, string> metadata = 9;
}

// DriftAck acknowledges receipt of a drift event
message DriftAck {
  string event_id = 1;
  bool received = 2;
  string error = 3;
}

// HealingResult reports the outcome of self-healing
message HealingResult {
  string agent_id = 1;
  string hostname = 2;
  string check_type = 3;
  bool success = 4;
  string error_message = 5;
  int64 timestamp = 6;
  map<string, string> artifacts = 7;  // e.g., bitlocker recovery key
}

// HealingAck acknowledges healing result
message HealingAck {
  string event_id = 1;
  bool received = 2;
}

// HeartbeatRequest is sent periodically to indicate liveness
message HeartbeatRequest {
  string agent_id = 1;
  int64 timestamp = 2;
  int32 drift_count = 3;      // Number of drift events since last heartbeat
  int32 healing_count = 4;    // Number of healing attempts
}

// HeartbeatResponse can signal config changes
message HeartbeatResponse {
  bool acknowledged = 1;
  bool config_changed = 2;    // If true, agent should re-register
}

// RMMStatusReport reports detected RMM tools
message RMMStatusReport {
  string agent_id = 1;
  string hostname = 2;
  repeated RMMAgent detected_agents = 3;
  int64 timestamp = 4;
}

// RMMAgent represents a detected RMM tool
message RMMAgent {
  string name = 1;           // e.g., "ConnectWise", "Datto", "NinjaRMM"
  string version = 2;
  bool running = 3;
  string service_name = 4;
}

// RMMAck acknowledges RMM status report
message RMMAck {
  bool received = 1;
}
